XCVario Driver

iltis42 opened the pull request:
<!--

Thank you for your interest in contributing to XCSoar! Please read the
following information to make it easier for us to review your changes.

We appreciate if you make sure to:

  - Document the changes (in the NEWS.txt file, and the manual when relevant)
  - Enable maintainer edits[1] (in case we need to help with the PR)
  - Check your commits and their messages (so they're all tidy and coherent)

[1] https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->


Brief summary of the changes

New protocol "$PXCV" to support new device, the XCVario ( https://xcvario.de ) featuring vario, TE, airspeed, temperature, AHRS (roll/bank, 3D acceleration), and bidirectional exchange for cruise/vario mode, MC, Ballast, Bugs and QNH. 
The new protocol is implemented in here, https://github.com/iltis42/OpenIVario and shall be the default for new series of XCVario's for 2021.
The new protocol allows all the above features, what won't be possible otherwise, and has been derived as a cherry pic from elements of OpenVario, Eye-Sensorbox, Borgelt, and Cambridge protocols.


<!--
Please note that the commit messages is where detailed descriptions of the
changes should be made - this section is just for a brief summary/overview.
-->


Related issues and discussions
------------------------------

<!--
Please link any relevant issues or forum posts here, for reference.

If this PR resolves an existing issue, please write "Closes #1234" so that
the issue is closed automatically when this PR is merged.
-->

lordfolken commented:
It looks very clean. I'm just a bit sad about the fact that it's another protocol. You previously implemented openvario protocol in your device, as far as i can see the main thing missing would have been acceleration information. I'm sure that we could discuss this with the openvario guys: 

RRR.R = roll angle in degree related to earth system,
III.I = pitch angle related to earth system,
X.XX = acceleration in X-Axis (multiples of G),
Y.YY = acceleration in Y-Axis,
Z.ZZ = acceleration in Z-Axis,

Something along the line of:

$POV,A,RRR.R,III.I,X.XX,Y.YY,Z.ZZ*05

What made you decide to implement a new protocol?

lordfolken commented:
@MaxKellermann the driver actually registers as XCVario, only the github repo is called openIvario.

MaxKellermann commented:
The PR title was OpenIVario.....
The commit messages are bad, and there is no link the publically available protocol documentation.
And I don't understand why there are two commits.

iltis42 commented:
> What made you decide to implement a new protocol?
The decision was as the XCVario is mostly a different gadget that the OpenVario sensor board with different capabilities and also different needs, and beside the AHRS extension, there are other things like bidirectional sync of MC, Ballast, Bug's and MC value. 
OpenVario uses a separate hardware to provide those input's to XCSoar, hence XCVario coming with a rotary switch and a setup option for MC,Ballast,Bugs and QNH needed to combine Setup and Sensor information in one protocol, as Bluetooth RFCOMM will only allow to add one device with one Bluetooth connection, or connecting XCVario with a serial line, from XCVario side there is only one line to connect. 
This restriction does not apply for OpenVario hardware that features several connection's.

lordfolken commented:
@iltis42 i.c. your reasoning. So as @MaxKellermann mentioned, please make one commit, as each commit should be fully functional in itself.

ps: max i changed the pr title to openivario as i misread.

hjr commented:
Merged commits as requested.

iltis42 commented:
> @iltis42 i.c. your reasoning. So as @MaxKellermann mentioned, please make one commit, as each commit should be fully functional in itself.
> 
> Also a webpage explaining the protocol would be fantastic, so we have a spec that isn't the code.
> ps: max i changed the pr title to openivario as i misread.

https://xcvario.de/handbuch?preview_id=241&preview_nonce=e9c8cf38d5&preview=true#__RefHeading___Toc5243_1697588519

I have added this part in the handbook, would this be enough, or needed more info ? 
The source format is german, but you can hit "Translate" an press your flag what will provide a google translation. 
Yes the vario's project github page is labeled "OpenIVario", reason was it started with only OpenVario protocol support, what was fine at that point. With further demands like setup synchronisation and new gimmicks in hardware, this is no more so lucky now, thinking even to rename the repo to xcvario later.

It has already been made one commit. Thanks also to @hjr for merging and reusing his developemt env to provide this quickly: https://github.com/XCSoar/XCSoar/pull/496/files

PS: Here a short demo of AHRS with the XCSoar fork using XCVario driver:
https://www.youtube.com/watch?v=eeck3aCUoGc&t=8s

MaxKellermann commented:
The link says "Du bist leider nicht berechtigt, Entwürfe anzusehen."

MaxKellermann commented:
Also, the link should be in the code - and a PR should not have any description that, that belongs into the commit message.
All content on GitHub is lost eventually, only real git content is preserved forever.

iltis42 commented:
> The link says "Du bist leider nicht berechtigt, Entwürfe anzusehen."
Sorry wrong entry, this is the correct link, also updated above.
https://xcvario.de/handbuch#__RefHeading___Toc5243_1697588519

hjr commented:
Added the link to the web site to the source code.
